### DISCUSSION TITLE
Support for ordered sequences?

### DISCUSSION THREAD

**Discussion by tslocke** [vlcn-io/cr-sqlite#90] [Q&A] [answered]
...does cr-sqlite have any special support for them?

Can always use fractional indexing if not, but I was wondering if they are supported in a more high level way?

Thanks

**Comment by tantaman** [comment, marked as answer]
Not at the moment. You'll need to use fractional indexing as you suggested.

See the reply below for many gotchas of fractional indexing. 

tldr: `crsqlite` will incorporate first class support for fractional indices in Q1 2023.

**Reply by tantaman** [reply]
@tslocke, as an update -- we're going to be adding first class support for fractional indexing given there are some problematic edge cases

- running out of precision after 64 inserts in just the right location
- users picking the same fractional index and no longer being able to sort something between those items
- some solutions to fractional indexing will always interleave results, some will not but incur a cost at insertion time

we've been talking through these issues here: https://discord.com/channels/989870439897653248/1052527039523078174

And as another update -- I'll be integrating `yjs` support next year (likely Q2) so you can store rich text crdts.

We'd originally discussed adding diamond types (#65) but `yjs` might be the fastest path forward to having something that works and most people are already familiar with `yjs`.

**Reply by tantaman** [reply]
We've converged on some approaches in the discord. We also discovered that SQLite supports ordering by [recursive relationships in a performant manner](https://vlcn.io/blog/recursive-ordering-in-sqlite) which should make RGA ([and similar](https://madebyevan.com/algos/crdt-tree-based-indexing/)) pretty trivial to implement.

For the direction I think we'll end up going with fractional indices, you can see my thoughts here: https://vlcn.io/blog/fractional-indexing

Still noodling on recursive ordering but current thoughts are here: https://vlcn.io/blog/recursive-ordering

and will likely end up implementing something like https://madebyevan.com/algos/crdt-tree-based-indexing/ but with compression for runs to compactly store documents.

[mutable tree hierarchy](https://madebyevan.com/algos/crdt-mutable-tree-hierarchy/) would be interesting but not something I have time to tackle at the moment.

**Comment by tantaman** [comment]
For sequences: going to take the simple route and just write a `yjs` provider that integrates with `crsqlite` clocks.

I.e., something just like https://github.com/yjs/y-leveldb but targeting sqlite.

We can continue rabbit-holing on more exotic options (https://github.com/vlcn-io/cr-sqlite/issues/65) in the future.
